### PR TITLE
PRL5230-hearing screen

### PR DESCRIPTION
**Pull request by sealDebolina**
**Before creating a pull request make sure that:**

- [ ] commit messages are meaningful and follow good commit message guidelines
- [ ] README and other documentation has been updated / added (if needed)
- [ ] tests have been updated / new tests has been added (if needed)
- [ ] the enable_keep_helm label has been added, if the helm release should be persisted after a successful build

Please remove this line and everything above and fill the following sections:

### Change description ###

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[ ] No
```

### CODE REVIEW

File: src/main/steps/common/yourhearings/hearings/template.njk [hunk 1]
```diff
@@ -374,6 +374,8 @@
    {% endfor %}
   {% endif %}
   {% if completedHearings.length>=1 %}
+  {# ...
```
Comment (vivek-sekhar) [inline src/main/steps/common/yourhearings/hearings/template.njk:377]: Please handle all hearing methods in util.ts

File: src/main/steps/common/yourhearings/hearings/content.ts [hunk 2]
```diff
@@ -320,6 +322,41 @@ export function getHearingMethod(req: AppRequest, attendees: Attendee[]): string
...
+    //   value:[HEARING_METHOD.VID,HEARING_METHOD.VIDCVP,HEARING_METHOD.VIDOTHER,HEARING_METHOD.VIDPVL,HEARING_METHOD.VIDSKYPE,HEARING_METHOD.VIDTEAMS,HEARING_METHOD.VIDVHS]
+    // },
+    // {
+    //   type:"TEL",
+    //   value:[HEARING_METHOD.TEL,HEARING_METHOD.TELBTM,HEARING_METHOD.TELCVP,HEARING_METHOD.TELOTHER,HEARING_METHOD.TELSKYP,HEARING_METHOD.TELTEMP]
+    // }]
+    // hearingMethod= method.map(hm => hm.value.find(i=>i==hearingMethod)?hm.type:hearingMethod)
+    // return hearingMethod[0];
+
+    // My suggestion //
```
Comment (vivek-sekhar) [inline src/main/steps/common/yourhearings/hearings/content.ts:336]: Refactor it and use less complex solutions as discussed.

**Review by vivek-sekhar — CHANGES_REQUESTED** [review]

File: src/main/steps/common/yourhearings/hearings/utils.ts [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+import dayjs from 'dayjs';
+
+import { Attendee, CaseType, HearingsList, PartyType, hearingDay } from '../../../../app/case/definition';
+import { AppRequest } from '../../../../app/controller/AppRequest';
+import { HEARING_METHOD } from '../../../../steps/constants';
+import { getCasePartyType } from '../../../../steps/prl-cases/dashboard/utils';
+
+import { cy, en } from './content';
+
+export function getHearingMethod(req: AppRequest, attendees: Attendee[]): string {
```
Comment (vivek-sekhar) [inline src/main/steps/common/yourhearings/hearings/utils.ts:13]: change it to getHearingDetails

File: src/main/steps/common/yourhearings/hearings/utils.ts [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+  const partyType = getCasePartyType(req.session.userCase, req.session.user.id);
+  const caseType = req.session.userCase.caseTypeOfApplication;
+  let partyData;
+  let hearingMethod;
+  const lang = req.session.lang === 'cy' ? cy : en;
+  if (caseType === CaseType.C100) {
```
Comment (vivek-sekhar) [inline src/main/steps/common/yourhearings/hearings/utils.ts:19]: resuse getPartyDetails method which is already written

File: src/main/steps/common/yourhearings/hearings/utils.ts [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+        applicant => applicant?.value?.user?.idamId === req.session.user.id
+      );
+    }
+    hearingMethod = attendees.find(attendee => attendee.partyID === partyData.id)!.hearingSubChannel;
+  } else {
+    partyData =
+      partyType === PartyType.RESPONDENT ? req.session.userCase.respondentsFL401 : req.session.userCase.applicantsFL401;
+    hearingMethod = attendees.find(attendee => attendee.partyID === partyData.partyId)!.hearingSubChannel;
+  }
+  if (hearingMethod !== null) {
```
Comment (vivek-sekhar) [inline src/main/steps/common/yourhearings/hearings/utils.ts:35]: use simple check

File: src/main/steps/common/yourhearings/hearings/utils.ts [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+    const a = {
```
Comment (vivek-sekhar) [inline src/main/steps/common/yourhearings/hearings/utils.ts:36]: const a?

File: src/main/steps/common/yourhearings/hearings/utils.ts [hunk 7]
```diff
@@ -0,0 +1,168 @@
...
+      judgeName,
+      venue,
+      address,
+      roomId,
+      detailsBlock,
+    });
+  }
+};
+
+export function getProperTime(date: Date): string {
```
Comment (vivek-sekhar) [inline src/main/steps/common/yourhearings/hearings/utils.ts:117]: what does it mean by getProperTime. Rename it to generateHearingTimeDisplayText

File: src/main/steps/common/yourhearings/hearings/utils.ts [hunk 8]
```diff
@@ -0,0 +1,168 @@
...
+  }
+  return dateString;
+}
+
+export type Row = {
+  key: string;
+  value: string | null | undefined;
+};
+
+export const dateGenaration = (req: AppRequest<Partial<Case>>, endDate: string) => {
```
Comment (vivek-sekhar) [inline src/main/steps/common/yourhearings/hearings/utils.ts:134]: generateHearingDate

File: src/main/steps/common/yourhearings/hearings/utils.ts [hunk 9]
```diff
@@ -0,0 +1,168 @@
...
+  return req.session.lang === 'cy'
+    ? dayjs(endDate).locale('cy').format('DD MMMM YYYY').toString()
+    : dayjs(endDate).locale('en').format('DD MMMM YYYY').toString();
+};
+export const dateGenarationWithDay = (req: AppRequest<Partial<Case>>, date: string) => {
```
Comment (vivek-sekhar) [inline src/main/steps/common/yourhearings/hearings/utils.ts:139]: generateHearingDateDisplayText

File: src/main/steps/common/yourhearings/hearings/utils.ts [hunk 10]
```diff
@@ -0,0 +1,168 @@
...
+  return req.session.lang === 'cy'
+    ? dayjs(date).locale('cy').format('dddd, D MMMM YYYY')
+    : dayjs(date).locale('en').format('dddd, D MMMM YYYY');
+};
+
+export const generateHearingText = (durationInDayOrHours: number, minutes: number, lang) => {
```
Comment (vivek-sekhar) [inline src/main/steps/common/yourhearings/hearings/utils.ts:145]: generateHearingScheduleDisplayText

File: src/main/steps/common/yourhearings/hearings/utils.ts [hunk 11]
```diff
@@ -0,0 +1,168 @@
...
+        return `${hourText} ${minuteText}`;
+      } else {
+        return generateMinuteText(minutes, lang);
+      }
+    } else {
+      return generateHourText(durationInDayOrHours, lang);
+    }
+  }
+};
+export const generateMinuteText = (minutes: number, lang: any) => {
```
Comment (vivek-sekhar) [inline src/main/steps/common/yourhearings/hearings/utils.ts:162]: generateHearingMinuteDisplayText

File: src/main/steps/common/yourhearings/hearings/utils.ts [hunk 12]
```diff
@@ -0,0 +1,168 @@
...
+  return minutes === 1 ? `1 ${lang.minute}` : `${minutes} ${lang.minutes}`;
+};
+
+export const generateHourText = (durationInDayOrHours: number, lang: any) => {
```
Comment (vivek-sekhar) [inline src/main/steps/common/yourhearings/hearings/utils.ts:166]: generateHearingHourDisplayText

File: src/main/steps/respondent/yourhearings/hearings/content.ts [hunk 13]
```diff
@@ -1,13 +1,25 @@
...
+} from '../../../../steps/urls';
 import { generateContent as yourhearingshearingscontent } from '../../../common/yourhearings/hearings/content';
 
 export { form } from '../../../common/yourhearings/hearings/content';
 export const generateContent: TranslationFn = content => {
   const hearingsContent = yourhearingshearingscontent(content);
-  hearingsContent.linkforsupport = CA_DA_ATTENDING_THE_COURT;
   const request = content.additionalData?.req;
+  const caseData = request.session.userCase;
+  if (content.additionalData?.req.session.userCase.caseTypeOfApplication === CaseType.C100) {
```
Comment (vivek-sekhar) [inline src/main/steps/respondent/yourhearings/hearings/content.ts:18]: use caseData

Comment (vivek-sekhar) [reply src/main/steps/respondent/yourhearings/hearings/content.ts:18]: use ternary operator

**Review by vivek-sekhar — CHANGES_REQUESTED** [review]

File: src/main/steps/common/yourhearings/hearings/utils.ts (hunk 3 above)
Comment (sealDebolina) [reply src/main/steps/common/yourhearings/hearings/utils.ts:22]: this method is retuening method type like in person, audio, video, hence the name

File: src/main/steps/common/yourhearings/hearings/utils.ts (hunk 4 above)
Comment (sealDebolina) [reply src/main/steps/common/yourhearings/hearings/utils.ts:28]: As we have discussed earlier getPartyDetails method can't be reused here

File: src/main/app/case/definition.ts [hunk 14]
```diff
@@ -2453,18 +2454,22 @@ export interface hearingDay{
   hearingDate: string,
   startTime: string,
   amPm:string,
+  preparedStartTime:string,
```
Comment (vivek-sekhar) [inline src/main/app/case/definition.ts:2457]: startTimeDisplayText

File: src/main/app/case/definition.ts [hunk 15]
```diff
@@ -2453,18 +2454,22 @@ export interface hearingDay{
...
   durationInDayOrHours:number,
   minutes:number,
+  hearingDurationText: string,
```
Comment (vivek-sekhar) [inline src/main/app/case/definition.ts:2460]: hearingDurationDisplayText

File: src/main/app/case/definition.ts [hunk 16]
```diff
@@ -2453,18 +2454,22 @@ export interface hearingDay{
...
   judgeName:string | null | undefined,
   venue:string | null | undefined,
   address:string | null | undefined,
   roomId:string | null | undefined,
+  detailsBlock: any
```
Comment (vivek-sekhar) [inline src/main/app/case/definition.ts:2465]: hearingToAttendDetails and use appropriate type

File: src/main/app/case/definition.ts [hunk 17]
```diff
@@ -2453,18 +2454,22 @@ export interface hearingDay{
...
 }
 
 export interface CompletedHearings{
   hearingId: Number | undefined,
   dates: string,
   lengthOfHearing: number | undefined,
+  lengthOfHearingText: string,
```
Comment (vivek-sekhar) [inline src/main/app/case/definition.ts:2472]: hearingDurationDisplayText

File: src/main/steps/common/yourhearings/hearings/utils.ts [hunk 18]
```diff
@@ -0,0 +1,305 @@
...
+    const hearingDurationText = generateHearingScheduleDisplayText(durationInDayOrHours, minutes, lang);
+
+    const judgeName = schedule.hearingJudgeName;
+    const venue = schedule.hearingVenueName;
+    const address = schedule.hearingVenueAddress;
+    const roomId = schedule.hearingRoomId;
+    const detailsBlock: Row[] = [];
+    if (hearingMethod === lang.inter) {
+      detailsBlock.push(
+        { key: lang.venue, value: venue },
```
Comment (vivek-sekhar) [inline src/main/steps/common/yourhearings/hearings/utils.ts:102]: displayText

File: src/main/steps/common/yourhearings/hearings/utils.ts [hunk 19]
```diff
@@ -0,0 +1,305 @@
...
+  hearingDay,
+  hearingStatus,
+} from '../../../../app/case/definition';
+import { AppRequest } from '../../../../app/controller/AppRequest';
+import { HEARING_METHOD } from '../../../../steps/constants';
+import { getCasePartyType } from '../../../../steps/prl-cases/dashboard/utils';
+
+import { cy, en } from './content';
+
+export function getHearingMethod(req: AppRequest, attendees: Attendee[]): string {
```
Comment (vivek-sekhar) [inline src/main/steps/common/yourhearings/hearings/utils.ts:22]: pls use arrow fns

**Review by vivek-sekhar — CHANGES_REQUESTED** [review]

File: src/main/steps/common/yourhearings/hearings/template.njk (hunk 1 above)
Comment (sealDebolina) [reply src/main/steps/common/yourhearings/hearings/template.njk]: done

File: src/main/steps/common/yourhearings/hearings/content.ts (hunk 2 above)
Comment (sealDebolina) [reply src/main/steps/common/yourhearings/hearings/content.ts]: done

File: src/main/steps/common/yourhearings/hearings/utils.ts (hunk 5 above)
Comment (sealDebolina) [reply src/main/steps/common/yourhearings/hearings/utils.ts]: done

File: src/main/steps/common/yourhearings/hearings/utils.ts (hunk 7 above)
Comment (sealDebolina) [reply src/main/steps/common/yourhearings/hearings/utils.ts]: done

File: src/main/steps/common/yourhearings/hearings/utils.ts (hunk 9 above)
Comment (sealDebolina) [reply src/main/steps/common/yourhearings/hearings/utils.ts]: done

File: src/main/steps/common/yourhearings/hearings/utils.ts (hunk 10 above)
Comment (sealDebolina) [reply src/main/steps/common/yourhearings/hearings/utils.ts]: done

File: src/main/steps/common/yourhearings/hearings/utils.ts (hunk 12 above)
Comment (sealDebolina) [reply src/main/steps/common/yourhearings/hearings/utils.ts]: done

File: src/main/steps/common/yourhearings/hearings/utils.ts (hunk 11 above)
Comment (sealDebolina) [reply src/main/steps/common/yourhearings/hearings/utils.ts]: done

File: src/main/steps/respondent/yourhearings/hearings/content.ts (hunk 13 above)
Comment (sealDebolina) [reply src/main/steps/respondent/yourhearings/hearings/content.ts]: done

File: src/main/app/case/definition.ts (hunk 14 above)
Comment (sealDebolina) [reply src/main/app/case/definition.ts]: done

File: src/main/app/case/definition.ts (hunk 16 above)
Comment (sealDebolina) [reply src/main/app/case/definition.ts]: done

File: src/main/app/case/definition.ts (hunk 17 above)
Comment (sealDebolina) [reply src/main/app/case/definition.ts]: done

File: src/main/steps/common/yourhearings/hearings/utils.ts (hunk 18 above)
Comment (sealDebolina) [reply src/main/steps/common/yourhearings/hearings/utils.ts]: done

File: src/main/steps/common/yourhearings/hearings/utils.ts [hunk 20]
```diff
@@ -0,0 +1,305 @@
...
+};
+export const generateHearingMinuteDisplayText = (minutes: number, lang: any) => {
+  return minutes === 1 ? `1 ${lang.minute}` : `${minutes} ${lang.minutes}`;
+};
+
+export const generateHearingHourDisplayText = (durationInDayOrHours: number, lang: any) => {
+  return durationInDayOrHours === 1 ? `1 ${lang.hour}` : `${durationInDayOrHours} ${lang.hours}`;
+};
+
+export const prepareHearingData = (req: AppRequest<Partial<Case>>): any => {
```
[the remark this answers is not in the document]
Comment (sealDebolina) [reply src/main/steps/common/yourhearings/hearings/utils.ts]: done

File: src/main/steps/common/yourhearings/hearings/utils.ts (hunk 19 above)
Comment (sealDebolina) [reply src/main/steps/common/yourhearings/hearings/utils.ts]: done

**Review by vivek-sekhar — APPROVED** [review]

**Review by AlokDatta — APPROVED** [review]

**Pull request merged**